src/test/test-capability.c

```diff
@@ -77,6 +86,34 @@ static int setup_tests(void) {
         test_uid = nobody->pw_uid;
         test_gid = nobody->pw_gid;
 
+        /* Find out the kernel version to determine if the ambient
+         * capability tests should be run. */
+
+        if (uname(&u) < 0)
+                return -errno;
+
+        version = strdup(u.release);
+        orig = version;
+
+        token = strsep(&version, ".");
+        if (!token || !version) {
+                return -EINVAL;
+        }
+
+        r = safe_atou(token, &major);
+        if (r < 0)
+                return r;
+
+        /* Can't use safe_atou here because the string is not clean. */
+        minor = strtoul(version, &end, 10);
+        if (version == end) {
+                return -EINVAL;
+        }
+
+        /* Only run ambient capability tests on kernel versions >= 4.3. */
+        if (major > 4 || (major == 4 && minor >= 3))
+                *run_ambient = true;
+
```

Comment:
We could try to use `prctl()` to enable an ambient capability when the inherited capability set does not have that cap. Then we should get back `EPERM` if the feature is supported in kernel and `EINVAL` otherwise. I'm just bit hesitant towards this approach, because it is possible to get back `EINVAL` also for other reasons, such as trying to set a capability that doesn't exist.

As you say, the version check has the problem of skipping ambient capability tests if an old kernel has backported patches. The problem is likely to be limited to only few distros and be transient in nature as those distros update their kernels. Still, I don't have a strong opinion about this, and I can look into the `prctl()` approach if you feel it makes more sense.